sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ChannelEnumerableSubscription.cs

```diff
@@ -0,0 +1,255 @@
+        /// <param name="maximumWaitTime">The maximum amount of time to wait to for an event to be available before emitting an empty item; if <c>null</c>, empty items will not be emitted.</param>
+        /// <param name="random">The random number generator to use for calculations.</param>
+        /// <param name="cancellationToken">The <see cref="System.Threading.CancellationToken"/> instance to signal the request to cancel enumeration.</param>
+        ///
+        /// <returns>An asynchronous enumerator that can be used to iterate over events as they are available.</returns>
+        ///
+        private static async IAsyncEnumerable<T> EnumerateChannel(ChannelReader<T> reader,
+                                                                  TimeSpan? maximumWaitTime,
+                                                                  Random random,
+                                                                  CancellationToken cancellationToken)
+        {
+            int delayMilliseconds;
+            T result;
+
+            var waitTime = maximumWaitTime?.TotalMilliseconds;
+            var timer = (waitTime.HasValue) ? Stopwatch.StartNew() : default;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                // Attempt to read events from the channel.
+
+                if (reader.TryRead(out result))
+                {
+                    yield return result;
+                    timer?.Restart();
+                }
+                else if ((waitTime.HasValue) && (timer.ElapsedMilliseconds > waitTime.Value))
+                {
+                    // If there was no event and the wait time has expired, emit an empty
+                    // event to return control to the calling iterator.
+
+                    yield return default;
+                    timer.Restart();
+                }
+                else if (reader.Completion.IsCompleted)
+                {
+                    // If the channel was marked as final, then await the completion task
+                    // to surface any exceptions.
+
+                    try
+                    {
+                        await reader.Completion.ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        // This is an expected case when the cancellation token was
+                        // triggered during an operation; no action is needed.
+                    }
+
+                    break;
+                }
+                else
+                {
+                    try
+                    {
+                        // In the case where no events are available, introduce a small random delay to avoid
+                        // pumping the channel in a tight loop.  Ensure that the calculated delay respects the
+                        // maximum wait time, if it is shorter.
+
+                        delayMilliseconds = random.Next(50, 150);
```

Comment:
This would kill the perf. Situations where the channel is empty but an item would arrive very soon are frequent in hight load apps.

Reply:
Good thought.  This was a carry-over from the track one approach, but after looking between them, I think the channel offers us enough protection here.   I'm going to pull this and we can see if anything comes up in real world usage.

Reply:
Hm, I don't think it has enough protection if you don't use ReadAsync. Doing TryRead in a tight loop would burn the CPU.

Reply:
I'm remembering why I did it that way now.    I originally had been creating a new CancellationTokenSource linked to the main token each iteration that timed out after the wait time and calling `WaitForReadAsync`, but that seemed like a ton of allocations.  If there was only a way to reset the timer on a cancellation source or the source itself, but nope.   The approach that I took was potentially less efficient between batches of events but avoided the tight loop and did not require allocations.

I'm going to revert to the random approach for preview 2, unless you have an idea for an alternate way that doesn't involve removing the wait time pattern.

Reply:
`Task.Delay` is not free either, it has to create a timer internally. You can create a timer of your own and attach it to the cancellation source, every time you get a message you reset elapsed time. This way there is a constant amount of allocations that don't depend on received message count.

Reply:
Scratch that just use CancelAfter method on CancellationSource. It does exactly what I just wrote.

Reply:
Some inspiration struck me this morning based on your thoughts.  I'm going to need time to prototype and play with the idea a bit to refine it, but I'm out of time for the moment.   I'm going to keep the simple static delay approach in place for code complete and revisit after  _(hopefully early next week)_.  

Tracking this with:   [Revisit Async Iterator Loop](https://github.com/Azure/azure-sdk-for-net/issues/7094) (#7094)